Task/due date column

jduong09 opened the pull request:
Resolves #49

jduong09 commented:
- May need to refactor updateSubscription.js and createSubscription.js. They share identical functions.
- Will need to discuss with Dan about wording of labels and divs.

jduong09 commented:
Changing to draft due to non-merged branch. Will merge main into this branch, and then convert back to open.

jduong09 commented:
Merged main branch into task/dueDateColumn. Added error handling for dueDateColumn. Ready to be reviewed.

ziggysauce commented on client/src/components/subscription/createSubscription.js:
```diff
@@ -55,11 +69,88 @@ class CreateSubscription extends React.Component {
 
     this.handleSubscriptions(response);
-    this.setState({ name: '', nickname: '', dueDate: todaysDate, reminderDays: 0, amount: 0 });
+    this.setState({ name: '', nickname: '', reminderDays: 0, amount: 0, frequency: '', occurence: 0, days: []});
```
```suggestion
    this.setState({ name: '', nickname: '', reminderDays: 0, amount: 0, frequency: '', occurence: 0, days: [] });
```

ziggysauce commented on client/src/components/subscription/createSubscription.js:
```diff
@@ -55,11 +69,88 @@ class CreateSubscription extends React.Component {
+              </select>
+            </label>
+            <div>Day of the week:</div>
+            <ReactDayPicker disabledDays={{ after: new Date(2022, 0, 8), before: new Date(2022, 0, 2) }} handleUpdate={this.handleDays} resetDays="weekly" />
```
these dates seem hard-coded to the first week of january? maybe instead just grab today, then determine the `before` and `after` date bounds based off of that

jduong09 replied:
My thought process:

- We want to display 7 days for the user to choose, so then we can give them a reminder based on the day of the week. ex: [Sunday] || [Monday, Thursday]
- The change I'm pushing up will be to choose the first day of the month the user is in. Then allow them to choose a day from the 1st to the 7th. 
- My problem with choosing today, and then determining 7 days after, is that if it's towards the end of the month, the 7 days will leak into the next month, leading to user having to click the next month.
- Should we scrap the calendar on this weekly part of the form?

ziggysauce commented on client/src/components/subscription/subscription.js:
```diff
@@ -3,11 +3,13 @@ import React from 'react';
 const Subscription = ({ details, handleEdit, handleDelete }) => {
   const { name, nickname, dueDate, reminderDays, amount } = details;
 
+  // const datesList = dueDate.dates.map((day) => )
```
remove commented out code

ziggysauce commented on client/src/components/subscription/updateSubscription.js:
```diff
@@ -1,21 +1,40 @@
 import { toast } from 'react-toastify';
+import ReactDayPicker from '../utils/datepicker';
+
+const WEEKLY_DAYS = {
+  0: '2022-01-02T00:00:00',
+  1: '2022-01-03T00:00:00',
+  2: '2022-01-04T00:00:00',
+  3: '2022-01-05T00:00:00',
+  4: '2022-01-06T00:00:00',
+  5: '2022-01-07T00:00:00',
+  6: '2022-01-08T00:00:00',
+};
```
this is fine for now, but obviously in 2023 we'd want to update this. perhaps in the future we can update this to be dynamic, but i think it's fine for now

ziggysauce commented on client/src/components/subscription/updateSubscription.js:
```diff
@@ -64,9 +95,85 @@ class UpdateSubscription extends React.Component {
+    this.setState({ days });
+  }
+
+  parseDueDate() {
```
to your point, a lot of the logic in the update component is the same as the create component. what i think  we can do is abstract the "form" portion into a separate component, then just pass in the data if we're updating and keep the data empty/null if we're creating. i can help walk you through this if you need help

ziggysauce commented on client/src/components/utils/datepicker.js:
```diff
@@ -0,0 +1,36 @@
+  };
+
+  return (
+      <DayPicker onDayClick={handleDayClick} selectedDays={days} hideOnDayClick={false} disabledDays={disabledDays} />
```
```suggestion
    <DayPicker onDayClick={handleDayClick} selectedDays={days} hideOnDayClick={false} disabledDays={disabledDays} />
```

ziggysauce commented on server/sql/migrations/00006_alter_column_due_date.sql:
```diff
@@ -0,0 +1,7 @@
+ALTER TABLE subscriptions
+ALTER COLUMN due_date TYPE jsonb 
+USING jsonb_build_object(
+  'frequency', 'monthly',
+    'occurence', 1,
+    'dates', ARRAY[EXTRACT(DAY FROM due_date)]
+  );
```
```suggestion
USING jsonb_build_object(
  'frequency', 'monthly',
  'occurence', 1,
  'dates', ARRAY[EXTRACT(DAY FROM due_date)]
);
```

ziggysauce reviewed (changes requested):
code-wise this looks pretty good at first glance. when i get some more time i'll do a deeper dive and test it a bunch. the biggest thing I think you can do though is practice DRY (don't repeat yourself) code by making a component that houses the form logic for creating/updating a subscription so that the code just lives in one place. Then each individual `create` or `update` component can use that form component. This way you have all the business logic in one place, so updating it in the future will be easier to manage than having to change code in two places.

jduong09 commented:
I am not satisfied with my functions for parsing the dueDate object into a readable string, for displaying to the user. Any ideas would be appreciated.

ziggysauce commented on client/src/components/subscription/subscriptionForm.js:
```diff
@@ -0,0 +1,239 @@
+              <div>On which day(s) do you want to be reminded?</div>
+              <ReactDayPicker handleUpdate={this.handleDays} canChangeMonth={false} />
+          </div>
+        );
```
some indentation fixes here. also for `occurrence` you'll want to replace all instances to make sure spelling is correct

```suggestion
        return (
          <div>
            <label htmlFor="occurence">
              Every (?) Months:
              <input type="number" id="occurrence" value={occurrence} onChange={(event) => this.handleChange(event, 'occurrence')} min="0" max="12" />
            </label>
            <div>On which day(s) do you want to be reminded?</div>
            <ReactDayPicker handleUpdate={this.handleDays} canChangeMonth={false} />
          </div>
        );
```

ziggysauce commented on client/src/components/date/datepicker.js:
```diff
@@ -0,0 +1,41 @@
+const ReactDayPicker = ({ handleUpdate, disabledDays, canChangeMonth }) => {
+  const [days, setDate] = useState([]);
+
+  /*
```
Why is this commented out? And if we don't plan on using it, we can just remove it. If you are keeping this here for a specific reason, let's add some comments for why

ziggysauce commented on client/src/components/date/datepicker.js:
```diff
@@ -0,0 +1,41 @@
+  };
+
+  return (
+    <DayPicker onDayClick={handleDayClick} selectedDays={days} hideOnDayClick={false} disabledDays={disabledDays} canChangeMonth={canChangeMonth} />
```
When editing a subscription, the red circle shows the current date, but we're not displaying a previously selected date(s) by the user:

<img width="315" alt="Screen Shot 2022-02-12 at 1 25 31 PM" src="https://user-images.githubusercontent.com/25065329/153723535-0877ef79-36b5-4383-b1fd-a4e2c1d57994.png">

I think for subscription updates, we _should_ display the date they have currently selected so that they know what to switch it to (in this example my previous subscription has the 15th selected):

<img width="318" alt="Screen Shot 2022-02-12 at 1 25 39 PM" src="https://user-images.githubusercontent.com/25065329/153723549-21afd18e-8fc7-4cbf-9555-7d726a92bf13.png">

jduong09 replied:
https://github.com/jduong09/water-your-subs/pull/58

Check out this PR which should have this change.

ziggysauce commented on client/src/components/subscription/subscriptionForm.js:
```diff
@@ -0,0 +1,239 @@
+    super(props);
+
+    const { method, prevSubscription } = props;
+    let parseDate = prevSubscription ? prevSubscription.dueDate.dates : null;
```
you can use optional chaining here instead

```suggestion
    let parseDate = prevSubscription?.dueDate?.dates || null;
```

ziggysauce commented on client/src/components/subscription/subscriptionForm.js:
```diff
@@ -0,0 +1,239 @@
+      reminderDays: prevSubscription ? prevSubscription.reminderDays : 0,
+      amount: prevSubscription ? prevSubscription.amount/100 : 0,
+      frequency: prevSubscription ? prevSubscription.dueDate.frequency : '',
+      occurence: prevSubscription ? prevSubscription.dueDate.occurence : 1,
```
can use optional chaining on all of these instead of a ternary

ziggysauce commented on client/src/components/subscription/subscriptionForm.js:
```diff
@@ -0,0 +1,239 @@
+    const { checked, value } = event.target;
+    let updatedList = checkedDays;
+    if (checked) {
+      updatedList = [ ...checkedDays, value];
```
we should be consistent with spacing since our linter isn't catching this:

either use `[...checkedDays, value]` or `[ ...checkedDays, value ]`

ziggysauce replied:
this still needs addressing (I'm referring to the space at the end before the closing square bracket)

ziggysauce commented on client/src/components/subscription/subscriptionForm.js:
```diff
@@ -0,0 +1,239 @@
+    return (
+      <section>
+        <h2>{method === 'PUT' ? 'Create Subscription' : 'Update Subscription'}</h2>
+        <form onSubmit={this.handleSubmit}>
```
we can wait for the figma design, but this form can use some UI updating.

ziggysauce commented on client/src/utils/frontendDateUtils.js:
```diff
@@ -0,0 +1,79 @@
+
+    if (DateUtils.isSameDay(nearestDueDate, new Date(todaysDate))) {
+      nearestDueDate = new Date(todaysDate);
+      toast(`Your subcription ${name} is due!`);
```
```suggestion
      toast(`Your subscription ${name} is due!`);
```

ziggysauce replied:
though actually I think it may make more sense to word it as such: ``Your ${name} subscription is due!``

ziggysauce commented on client/src/utils/frontendDateUtils.js:
```diff
@@ -0,0 +1,79 @@
+    const dateObject = new Date(dates[i]);
+    if (DateUtils.isSameDay(dateObject, new Date(todaysDate))) {
+      nearestDueDate = dateObject;
+      toast(`Your subscription ${name} is due!`);
```
same comment as above

ziggysauce commented on client/src/utils/frontendDateUtils.js:
```diff
@@ -0,0 +1,79 @@
+    }
+  }
+
+  const dueDateString = `${nearestDueDate.toISOString().substring(0, 10)}, ${frequency}`;
```
So we save dates in the YYYY-MM-DD format, but for display purposes it may more sense to show it as MM/DD/YY (or even MM/DD) format

jduong09 replied:
Done in task/addNextDueDate PR.

ziggysauce commented on server/api/actions/subscriptions.js:
```diff
@@ -2,16 +2,16 @@ const db = require('../../db/db');
+      occurence: parseInt(requestBody.occurence),
       reminderDays: parseInt(requestBody.reminderDays),
-      amount: requestBody.amount * 100
+      amount: Math.round((requestBody.amount * 100) * 100) / 100
```
fine for now... but perhaps might make sense to create a utility function for normalizing the currency display value if this ends up being used more frequently

ziggysauce commented on server/api/routes/auth.js:
```diff
@@ -77,7 +77,6 @@ router.get('/callback', (req, res, next) => {
       // URGENT: Need to look at purpose of deleting returnTo
       delete req.session.returnTo;
       await res.redirect(`${process.env.BASE_URL}/users/${user_uuid}`);
-      res.end();
```
why did we remove this?

jduong09 replied:
I'm not sure. My mistake.

ziggysauce commented on server/api/routes/subscriptions.js:
```diff
@@ -48,4 +49,26 @@ router.delete('/:subscriptionUuid', async (req, res) => {
+
+    res.status(200).json({ lateDueDates });
+  } catch(error) {
+    console.log('Error: ', error);
```
we'll probably want to send a 5xx status code so that the client can handle the error

ziggysauce commented on server/utils/date.js:
```diff
@@ -0,0 +1,62 @@
+        return new Date(addDays(date, occurence * 7)).toISOString().substring(0, 10);
+      }
+    }
+    return day;
```
here you're returning a single date `YYYY-MM-DD` but below you're returning an object `{ name, dueDate }`. With functions like this you want to return a consistent data type. If they're not, then we should abstract separate functions to return the different data types

jduong09 replied:
I'm not sure I'm following this. Take a look at my changes and then please let me know if I've solved the problem.

ziggysauce commented on server/utils/date.js:
```diff
@@ -0,0 +1,62 @@
+
+  const body = {
+    dueDate: {
+      frequency, 
```
```suggestion
      frequency,
```

ziggysauce commented on server/utils/date.js:
```diff
@@ -0,0 +1,62 @@
+    }
+  }
+
+  const newDatesObject = await dates.map((day) => {
```
this can/should just be part of the if/else above to simplify logic. additionally, we can just combine some of the logic to make this all a bit easier to read and follow:

```javascript
  let lateDueDate;

  const formatDate = (frequency) => {
    let newDueDate;
    if (frequency === 'daily') {
      newDueDate = new Date(dates[0]);
      if (newDueDate < new Date(todaysDate)) {
        lateDueDate = newDueDate;

        while (newDueDate < new Date(todaysDate)) {
          newDueDate = new Date(addDays(newDueDate, parseInt(occurence, 10)));
        }
      }
      return [newDueDate.toISOString().substring(0, 10)];
    } else {
      const newDatesObject = dates.map((day) => {
        const date = new Date(day);
        if (date < new Date(todaysDate)) {
          lateDueDate = date.toISOString().substring(0, 10);

          if (frequency === 'yearly') {
            return addMonths(date, 12);
          } else if (frequency === 'monthly') {
            return addMonths(date, parseInt(occurence, 10));
          } else if (frequency === 'weekly') {
            return new Date(addDays(date, occurence * 7)).toISOString().substring(0, 10);
          }
        }
        return day;
      });
      return newDatesObject;
    }
  }

  const body = {
    dueDate: {
      frequency, 
      occurence,
      dates: formatDate(frequency),
      lateDueDate,
    },
    subscriptionUuid
  };
```

ziggysauce commented on server/utils/date.js:
```diff
@@ -0,0 +1,62 @@
+    subscriptionUuid
+  };
+
+  if (lateDueDate) {
+    body.dueDate.lateDueDate = lateDueDate;
+  }
```
no need to separate this, see comment above

ziggysauce commented on server/utils/date.js:
```diff
@@ -0,0 +1,62 @@
+    dueDate: {
+      frequency, 
+      occurence,
+      dates: (frequency === 'daily') ? [newDueDate.toISOString().substring(0, 10)] : newDatesObject
```
see comment above about simplifying this

ziggysauce commented on server/api/routes/subscriptions.js:
```diff
@@ -48,4 +49,26 @@ router.delete('/:subscriptionUuid', async (req, res) => {
+  const lateDueDates = [];
+  try {
+    const allSubscriptions = await getSubscriptionsByUserId(user_id);
+    for (let i = 0; i < allSubscriptions.length; i += 1) {
+      const subscription = allSubscriptions[i];
+      const updatedSubscription = await updateNextDueDate(subscription.dueDate, subscription.subscriptionUuid);
+      const { dueDate, name } = updatedSubscription;
+      if (updatedSubscription) {
+        if (dueDate.lateDueDate) {
+          lateDueDates.push({ name, date: dueDate.lateDueDate });
+        }
+      }
+    }
```
this may be worth doing in parallel since you're just updating each subscription individually and doesn't rely on other subscriptions finishing their update first. You can use `await Promise.all()` and then create an array of all of the updates then filter out those that don't have a lateDueDate.

another option I explored is the use `await Promise.all()` and then a `reduce()` in order to just pass back the updated data:

```javascript
  const { user_id } = req.session.userInfo;
  try {
    const allSubscriptions = await getSubscriptionsByUserId(user_id);
    const updatedSubscriptions = await Promise.all(allSubscriptions.map(({dueDate, subscriptionUuid}) =>
      updateNextDueDate(dueDate, subscriptionUuid)));

    const lateDueDates = updatedSubscriptions.reduce((lateSubs, subscription) => {
      if (subscription?.dueDate?.lateDueDate) {
        lateSubs.push({ name: subscription.name, date: subscription.dueDate.lateDueDate});
      }
      return lateSubs;
    }, []);

    res.status(200).json({ lateDueDates });
  } catch(error) {
    console.log('Error: ', error);
  }
```

ziggysauce commented on client/src/components/user/dashboard.js:
```diff
@@ -12,10 +13,32 @@ class Dashboard extends React.Component {
+
+      for (let i = 0; i < lateDueDates.length; i += 1) {
+        const { name, date } = lateDueDates[i];
+        toast(`Your subscription ${name} was due ${new Date(date).toISOString().substring(0, 10)}`);
```
is there a way to make the toast notification sticky and only go away when the user manually closes it? For subscriptions that are either past due or due today, I think the toast notification should stay on the page in case the user were to miss it.

Alternatively/additionally we could visually highlight the subscription cards for those that are either due today or past due with a simple border color difference or something else that stands out to the user

ziggysauce commented on client/src/components/date/datepicker.js:
```diff
@@ -0,0 +1,41 @@
+  }, [resetDays, updating]);
+  */
+  
+  const handleDayClick = (day, { selected }) => {
```
Do we want to disallow the user from selecting a date in the past? Right now we don't but since the "next due date" always just overrides the displayable "Due Date" I think it's worth just restricting the user from selecting a date/dates in the past

jduong09 replied:
I choose to allow the user to select days from the past. For selecting days of the month, I wanted them to be able to select any day in the current month, instead of having to select to go to the next month, and then choose a day in the month. Also, if it was the last day of the month, and they see a calendar that has one day available, and the other days disabled, it doesn't look that good to me. Let me know what you think.

ziggysauce replied:
I see. Yeah this is fine I guess, no biggie

ziggysauce reviewed (changes requested):
Awesomeeeee work here so far!  This is huge and works pretty well! A couple of suggestions here and there after testing and review the code. Can do a follow up review after changes are made.

ziggysauce commented on client/src/components/subscription/subscriptionForm.js:
```diff
@@ -0,0 +1,241 @@
+
+function convertStringToDate(datesArray) {
+  return datesArray.map((date) => new Date(date));
+};
```
can one-liner this fxn using a `const`

```suggestion
const convertStringToDate = (datesArray) => datesArray.map((date) => new Date(date));
```

ziggysauce reviewed (changes requested):
- Right now the calendar UI lets me select multiple dates. We probably only want to allow selecting one date unless we're explicitly allowing multi-date selection for the month view.

- When the subscription is due (current day) the toast notification remains which is good. But because it's white, it's a little easy to miss. I'd add some color (green?) to it so it stands out a bit.

- When the subscription is late (previous day) the toast notification still runs on a timer and disappears when I think we should also have it remain, and add some color, but maybe red to differentiate from the current due subscription UI.

- When I select no date at all from the calendar, it still lets me submit the form but shows an error and an invalid date in the UI. We should disallow the user from being able to submit if the date isn't selected. 
<img width="260" alt="Screen Shot 2022-02-19 at 5 56 32 PM" src="https://user-images.githubusercontent.com/25065329/154821899-b2f65c30-4cf6-4cd1-b7eb-a1748c057af0.png">

- When selecting a weekly subscription, we probably don't need to show the calendar, right? Right now it shows it, and when I try to select any date it blasts me with this error:
<img width="1297" alt="Screen Shot 2022-02-19 at 5 58 35 PM" src="https://user-images.githubusercontent.com/25065329/154821971-5f134bf0-2e43-4d04-92aa-b662c3b815ee.png">

- Furthermore, when I go to **edit** my weekly subscription all the checkboxes are empty, but we should display which day(s) of the week they previously selected so they know what to change it to.

- When I go to edit subscriptions and reload the page, the order that my subscription cards are in jump around. We should set an order in the SQL query that fetches all the subscriptions so that it has an `ORDER BY` sorting

jduong09 commented:
### Changes/Additions
- Removed the ability to select multiple dates, besides on <monthly> subscriptions. For yearly and daily subscriptions, disabled days after user clicks on a day on the calendar. If they try to click another day, display error notification that they cannot. 

- When subscription is due: Changed the background color and text color to match design. 

- When subscription is late: Removed timer, so it does not disappear, and only disappears when user clicks on notification. Changed background color to red, to signify that user missed a subscription date. 

- Fixed error where user would be able to submit form without selecting a date. If user tries to submit the form without a date selected, sends toast notification to alert them to pick a date.

- Kept calendar for weekly subscriptions, but made non-interactive. My thought process is that the user can see the date that they select, when they choose 'Monday' for when their weekly subscriptions.

- Checked the correct past weekday boxes for when user wants to update their subscription. Now, user can see which day (s) they had selected previously.

- Added an <Order By> clause to sql query in order to keep subscriptions from jumping around when user makes changes to their subscription list.

### Comments
- When a subscription is due, if they make changes to the subscription, but not to the due date, then the toast notification alerting them that it is due will stack twice. Need a way to reset all the toast notifications/delete them after user makes changes.
- Need to design toast notifications to fit how we want the website to look.
- Change form label for "Reminder Days"; Is there a better word we can use to describe this column?

ziggysauce commented on client/src/components/date/datepicker.js:
```diff
@@ -0,0 +1,62 @@
+      const selectedIndex = selectedDays.findIndex(selectedDay => DateUtils.isSameDay(selectedDay, day));
+      selectedDays.splice(selectedIndex, 1);
+      setDisabledDay(false);
+    } else  if ((frequency === 'yearly' || frequency === 'daily') && (days.length !== 0)) {
```
```suggestion
    } else if ((frequency === 'yearly' || frequency === 'daily') && (days.length !== 0)) {
```

ziggysauce replied:
last conditional doesn't need parens. AND I think you can just do `days.length` since 0 is falsey unless we ever get a negative number...?

```suggestion
    } else  if ((frequency === 'yearly' || frequency === 'daily') && days.length) {
```

ziggysauce commented on client/src/components/subscription/subscriptionForm.js:
```diff
@@ -0,0 +1,230 @@
+
+class SubscriptionForm extends React.Component {
+  constructor(props) {
+    super(props);  
```
```suggestion
    super(props);
```

jduong09 commented:
## Fixes
- Combined check for late and same day subscriptions into same function. If a subscription was late, then it will not display that it is also due today. 
- Removed the check to see if it is due today from <subscription.js>. Now, if a user creates a new subscription and sets the due date to today, There will not be a notification. Instead, if they refresh the page, they will then see that it is due today. This is due to the check for late and today subscriptions being in <dashboard.js>.
- Fixed a bug where clicking a weekly checkbox would display the day before the correct day. It should display the correct day. 
- In subscription.js, checked to see if the subscription has nickname, and if it does, then display nickname instead of name.

ziggysauce reviewed (changes requested):
I have this sub set for monthly (in this case Feb 18), but when I go to the landing page after that date it correctly updates the date on the subscription card but displays the incorrect due date in the error notification

<img width="1074" alt="Screen Shot 2022-02-21 at 10 42 04 PM" src="https://user-images.githubusercontent.com/25065329/155059234-19ce1639-74b9-46d5-a7cb-76c52a8d36b0.png">

ziggysauce reviewed (changes requested):
When I try to create a new weekly sub with "Sunday" as the selected day, I get hit with this:

<img width="1055" alt="Screen Shot 2022-02-21 at 10 48 10 PM" src="https://user-images.githubusercontent.com/25065329/155059675-f9ecb329-afb4-4183-b96c-d5e2fc4f0536.png">

jduong09 commented:
### Fixes
- Fixed issue occurring where app would display the late due date instead of the next due date, due to how componentDidMount and rendering works in React.
- Fixed issue where toast notification for late subscriptions would display the date as a day behind than the correct date.
- Fixed issue where selecting a day of the week on the weekly subscription form would cause app to crash.

### Tests
- Tested that user can select <only one> date on yearly and daily subscriptions.
- Tested that <multiple dates> are allowed on their respective subscription form.
- Tested that setting a subscription that is due in the past, would update to the correct date in the future, while displaying the correct toast notification.
- Tested that setting a subscription to todays date, updates to show next due date to today. Toast Notification will show on browser refresh.
- Tested that not setting a date causes an error, and form cannot be submitted.
- Tested multiple cases for when date is late.

ziggysauce commented on client/src/components/subscription/subscriptionsList.js:
```diff
@@ -34,15 +36,37 @@ class SubscriptionsList extends React.Component {
+      const { dueDate, name } = subscription;
+
+      if (dueDate.lateDueDate) {
+        toast.error(`Your subscription ${name} was due ${new Date(dueDate.lateDueDate).toLocaleDateString()}`, {
```
let's change the order of text so that it says `Your ${sub_name} subscription was due on ${date}` to match the "due today" toast noti

<img width="334" alt="Screen Shot 2022-02-26 at 4 03 07 PM" src="https://user-images.githubusercontent.com/25065329/155858980-2cf260ca-317d-4e5a-bcef-00d1539ff21f.png">

ziggysauce commented on client/src/components/subscription/subscriptionsList.js:
```diff
@@ -34,15 +36,37 @@ class SubscriptionsList extends React.Component {
+          style: { backgroundColor: 'red', color: '#000000' }
+        });
+      } else if (DateUtils.isSameDay(new Date(dueDate.nextDueDate), new Date()) && !dueDate.lateDueDate) {
+        toast(`Your ${name} subscription is due!`, {
```
```suggestion
        toast(`Your ${name} subscription is due today!`, {
```

ziggysauce reviewed (changes requested):
Awesome job! Changes look excellent. Last few minor things and this will be ready!

- When I have a weekly sub every 2 weeks (but applies for 3 or 4 weeks as well) the sub i think should say "Every x weeks" instead of right now just saying "Weekly"
<img width="265" alt="Screen Shot 2022-02-26 at 3 56 47 PM" src="https://user-images.githubusercontent.com/25065329/155858847-6a8c99b8-8e9b-4d91-8d2d-5f9e211f98ef.png">
<img width="1146" alt="Screen Shot 2022-02-26 at 3 56 52 PM" src="https://user-images.githubusercontent.com/25065329/155858849-6e84ffeb-acb3-49de-a418-642dff957744.png">

- Same as above with monthly when I select multiple dates. Instead of just saying "monthly" we should be specific so it's clear from the top-view when it's due. Here's how Apple ToDo does it and I think it works super well
![IMG_8241](https://user-images.githubusercontent.com/25065329/155858936-6bf6d64e-0f9d-4e11-bc39-ce5b68e51136.jpg)

jduong09 commented:
### Changes/Additions
- Added functions to parse the dueDate object into a readable sentence.
- Removed option to change months when user is making a monthly subscription and chose a date already.
- Changed toast notifications for late due dates and ones that are due day of.

ziggysauce commented on client/src/components/subscription/subscriptionsList.js:
```diff
@@ -34,15 +36,37 @@ class SubscriptionsList extends React.Component {
+      const { dueDate, name } = subscription;
+
+      if (dueDate.lateDueDate) {
+        toast.error(`Your subscription ${name} was due on ${new Date(dueDate.lateDueDate).toLocaleDateString()}`, {
```
to match below:


```suggestion
        toast.error(`Your ${name} subscription was due on ${new Date(dueDate.lateDueDate).toLocaleDateString()}`, {
```

ziggysauce reviewed (approved):
Looks good to me functionally! nice work on this man, this is tremendous!
